x509/src/general_name.rs

```diff
@@ -0,0 +1,417 @@
+///         dNSName                         [2]     IA5String,
+///         x400Address                     [3]     ORAddress,
+///         directoryName                   [4]     Name,
+///         ediPartyName                    [5]     EDIPartyName,
+///         uniformResourceIdentifier       [6]     IA5String,
+///         iPAddress                       [7]     OCTET STRING,
+///         registeredID                    [8]     OBJECT IDENTIFIER }
+/// ```
+///
+/// [RFC 5280 Section 4.2.1.6]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum GeneralName<'a> {
+    /// otherName                       [0]     OtherName,
+    OtherName(OtherName<'a>),
+
+    /// rfc822Name                      [1]     IA5String,
+    Rfc822Name(Ia5String<'a>),
+
+    /// dNSName                         [2]     IA5String,
+    DnsName(Ia5String<'a>),
+
+    // x400Address                     [3]     ORAddress,
+    // Not supporting x400Address
+    /// directoryName                   [4]     Name,
+    DirectoryName(Name<'a>),
+
+    // ediPartyName                    [5]     EDIPartyName,
+    // Not supporting ediPartyName
+    /// uniformResourceIdentifier       [6]     IA5String,
+    UniformResourceIdentifier(Ia5String<'a>),
+
+    /// iPAddress                       [7]     OCTET STRING,
+    IpAddress(OctetString<'a>),
+
+    /// registeredID                    [8]     OBJECT IDENTIFIER
+    RegisteredId(ObjectIdentifier),
+}
+
+const OTHER_NAME_TAG: TagNumber = TagNumber::new(0);
+const RFC822_NAME_TAG: TagNumber = TagNumber::new(1);
+const DNS_NAME_TAG: TagNumber = TagNumber::new(2);
+//const X400_ADDRESS_TAG: TagNumber = TagNumber::new(3);
+const DIRECTORY_NAME_TAG: TagNumber = TagNumber::new(4);
+//const EDIPI_PARTY_NAME_TAG: TagNumber = TagNumber::new(5);
+const URI_TAG: TagNumber = TagNumber::new(6);
+const IP_ADDRESS_TAG: TagNumber = TagNumber::new(7);
+const REGISTERED_ID_TAG: TagNumber = TagNumber::new(8);
+
+// Custom Decodable to handle implicit context specific fields (this may move to derived later).
+impl<'a> Decodable<'a> for GeneralName<'a> {
+    fn decode(decoder: &mut Decoder<'a>) -> der::Result<Self> {
+        let t = decoder.peek_tag()?;
+        match t.octet() {
+            0xA0 => {
+                let on =
+                    decoder.context_specific::<OtherName<'_>>(OTHER_NAME_TAG, TagMode::Implicit)?;
+                Ok(GeneralName::OtherName(on.unwrap()))
+            }
+            0x81 => {
+                let ia5 = decoder
+                    .context_specific::<Ia5String<'_>>(RFC822_NAME_TAG, TagMode::Implicit)?;
+                Ok(GeneralName::Rfc822Name(ia5.unwrap()))
+            }
+            0x82 => {
+                let ia5 =
+                    decoder.context_specific::<Ia5String<'_>>(DNS_NAME_TAG, TagMode::Implicit)?;
+                Ok(GeneralName::DnsName(ia5.unwrap()))
+            }
+            0xA3 => unimplemented!(),
```

Comment:
Are these a TODO? I think it's bad to panic as opposed to returning an error.

Reply:
Bah. Yes, those should be errors. I see no reason to mark as TODO but forgot to fix that. FIxed now.